testenv/aws/utils.go

```diff
@@ -0,0 +1,291 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/aws/aws-sdk-go-v2/service/cloudformation"
+	cloudformationtypes "github.com/aws/aws-sdk-go-v2/service/cloudformation/types"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
+
+	"github.com/openclarity/vmclarity/installation"
+	"github.com/openclarity/vmclarity/testenv/utils"
+	dockerhelper "github.com/openclarity/vmclarity/testenv/utils/docker"
+)
+
+func (e *AWSEnv) prepareStack(ctx context.Context) error {
+	var err error
+
+	// Create a new key pair
+	_, err = e.ec2Client.ImportKeyPair(ctx, &ec2.ImportKeyPairInput{
+		KeyName:           &e.stackName,
+		PublicKeyMaterial: e.sshKeyPair.PublicKey,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to import key pair: %w", err)
+	}
+
+	// Create a new S3 bucket
+	_, err = e.s3Client.CreateBucket(ctx, &s3.CreateBucketInput{
+		Bucket: &e.stackName,
+		CreateBucketConfiguration: &s3types.CreateBucketConfiguration{
+			LocationConstraint: s3types.BucketLocationConstraint(e.region),
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create bucket: %w", err)
+	}
+
+	f, err := installation.AWSManifestBundle.Open("VmClarity.cfn")
+	if err != nil {
+		return fmt.Errorf("failed to read template file: %w", err)
+	}
+
+	// Upload template file to S3 bucket
+	_, err = e.s3Client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket: &e.stackName,
+		Key:    &e.stackName,
+		Body:   f,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to put object: %w", err)
+	}
+	e.templateURL = "https://" + e.stackName + ".s3.amazonaws.com/" + e.stackName
+
+	return nil
+}
+
+func (e *AWSEnv) afterSetUp(ctx context.Context) error {
+	ready, err := e.infrastructureReady(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to check if infrastructure is ready: %w", err)
+	}
+	if !ready {
+		return errors.New("infrastructure is not ready")
+	}
+
+	if err = e.server.WaitForSSH(ctx, DefaultSSHPortReadyTimeout); err != nil {
+		return fmt.Errorf("failed to wait for the SSH port to become ready: %w", err)
+	}
+
+	e.sshPortForwardInput = &utils.SSHForwardInput{
+		PrivateKey:    e.sshKeyPair.PrivateKey,
+		User:          DefaultRemoteUser,
+		Host:          e.server.PublicIP,
+		Port:          utils.DefaultSSHPort,
+		LocalPort:     8080, //nolint:gomnd
+		RemoteAddress: "localhost",
+		RemotePort:    80, //nolint:gomnd
+	}
+
+	sshPF, err := utils.NewSSHPortForward(e.sshPortForwardInput)
+	if err != nil {
+		return fmt.Errorf("failed to setup SSH port forwarding: %w", err)
+	}
+
+	// Use non-inherited context to avoid cancelling the port forward with timeout
+	if err = sshPF.Start(context.Background()); err != nil { //nolint:contextcheck
```

Comment:
if this is the case, port forward will remain running? should we have a context ontop of the AWSEnv passed from testenv to ensure that port forwarding stops after the tests have been executed?

Reply:
another note: ssh port forwarding is never stopped (either explicitly via stop, or via context), which is needed as otherwise the port will stay in use